packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+            // TODO: can we get away with only marking the owner as exported, and not everything up to the package?
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+
+            if (sym.data(ctx)->flags.isStaticFieldTypeAlias) {
+                if (auto e = ctx.beginError(send.loc, core::errors::Packager::ExportingTypeAlias)) {
+                    e.setHeader("Exporting a type alias is not allowed in package specifications");
+                    e.addErrorLine(sym.data(ctx)->loc(), "Originally defined here");
+                }
+            }
+        }
+
+        return tree;
+    }
+
+    ast::ExpressionPtr preTransformClassDef(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+
+        if (original.symbol == core::Symbols::root()) {
+            return tree;
+        }
+
+        setPackageLocs(ctx, original.name.loc(), original.symbol);
+
+        return tree;
+    }
+
+    static ast::ParsedFile run(core::GlobalState &gs, ast::ParsedFile f) {
+        if (!f.file.data(gs).isPackage()) {
+            return f;
+        }
+
+        auto pkgName = gs.packageDB().getPackageNameForFile(f.file);
+        if (!pkgName.exists()) {
+            return f;
+        }
+
+        const auto &package = gs.packageDB().getPackageInfo(pkgName);
+        if (!package.exists()) {
+            return f;
+        }
+
+        PropagateVisibility pass{package};
+
+        pass.exportPackageRoots(gs);
+
+        core::MutableContext ctx{gs, core::Symbols::root(), f.file};
+        f.tree = ast::TreeMap::apply(ctx, pass, std::move(f.tree));
+
+        return f;
+    }
+};
+
+class VisibilityCheckerPass final {
+    bool nameMatchesPackage(const std::vector<core::NameRef> &revNameParts) {
+        const auto &packageName = this->package.fullName();
+
+        if (revNameParts.size() < packageName.size()) {
+            return false;
+        }
+
+        return std::equal(packageName.begin(), packageName.end(), revNameParts.rbegin());
+    }
+
+public:
+    const core::packages::PackageInfo &package;
+    const bool insideTestFile;
+
+    VisibilityCheckerPass(core::Context ctx, const core::packages::PackageInfo &package)
+        : package{package}, insideTestFile{ctx.file.data(ctx).isPackagedTest()} {}
+
+    // `keep-def` will reference constants in a way that looks like a packaging violation, but is actually fine. This
+    // boolean allows for an early exit when we know we're in the context of processing one of these sends. Currently
+    // the only sends that we process this way will not have any nested method calls, but if that changes this will need
+    // to become a stack.
+    bool ignoreConstant = false;
+
+    ast::ExpressionPtr preTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        this->ignoreConstant = send.fun == core::Names::keepForIde();
```

Comment:
```suggestion
        ENFORCE(!this->ignoreConstant, "keepForIde has nested sends");
        this->ignoreConstant = send.fun == core::Names::keepForIde();
```

This suitably enforces the comment above, right?